File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.py
class Solution(object):
    def minExtraChar(self, s, dictionary):
        sz = len(s)
        dp = [sz]*(sz+1)
        se = set(dictionary)
        dp[0]=0
        for i in range(1,sz+1):
            for j in range(i):
                temp=s[j:i]
                if temp in se:
                    dp[i]=min(dp[i],dp[j])
            dp[i]=min(dp[i],dp[i-1]+1)
            
        return dp[sz]
        """
        :type s: str
        :type dictionary: List[str]
        :rtype: int
        """
        

File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.cpp
class Solution {
public:
    int minExtraChar(string s, vector<string>& dictionary) {
        int sz=s.size();
        vector<int>dp(sz+1,sz);
        unordered_set<string>se;
        se.insert(dictionary.begin(),dictionary.end());
        dp[0]=0;
        int res=sz;
        for(int i=1;i<=sz;i++){
            for(int j=0;j<i;j++){
                string temp=s.substr(j,i-j);
                if(se.find(temp)!=se.end()){
                    dp[i]=min(dp[i],dp[j]);
                }
            }
            dp[i]=min(dp[i],dp[i-1]+1);
        }
        return dp[sz];
    }
};